include/emitc/emitc_tosa.h

```diff
@@ -137,6 +139,120 @@ Dest fully_connected(Src input, Weights weights, Bias bias) {
   return output;
 }
 
+/// Reduce ops
+// ReduceOp
+template <typename Dest, typename Src, typename Computation>
+inline Dest reduce(Src operand, typename get_element_type<Src>::type initValue,
+                   int64_t dimension, Computation computation) {
+  static_assert(is_tensor<Src>::value, "Expected tensor argument");
+  static_assert(is_tensor<Dest>::value, "Expected tensor result");
+
+  using ET_Src = typename get_element_type<Src>::type;
+  using ET_Dest = typename get_element_type<Src>::type;
+
+  static_assert(std::is_same<ET_Src, ET_Dest>::value, "Element type mismatch");
+
+  static_assert(Src::rank() == Dest::rank() + 1,
+                "source rank must equal dest rank + 1");
+
+  std::vector<size_t> retainedDimensions(Src::rank());
+  std::iota(retainedDimensions.begin(), retainedDimensions.end(), 0);
+  retainedDimensions.erase(retainedDimensions.begin() + dimension);
+
+  assert(retainedDimensions.size() == Dest::rank());
+
+  Dest result;
+  std::fill(result.begin(), result.end(), initValue);
+
+  for (size_t i = 0; i < operand.size(); i++) {
+    auto value = operand[i];
+    auto index = operand.unravel_index(i);
+
+    std::array<size_t, Dest::rank()> reducedIndex;
+    size_t j = 0;
+    for (size_t dim : retainedDimensions) {
+      reducedIndex[j++] = index[dim];
+    }
+
+    auto reductionValue = result[result.ravel_index(reducedIndex)];
+    result[result.ravel_index(reducedIndex)] =
+        computation(reductionValue, value);
+  }
+
+  return result;
+}
```

Comment:
There is no `tosa.reduce`, so I think this has to be a core op or if it isn't it should be implemented in an anonymous namespace?

Reply:
I put `reduce` in an anonymous namespace :+1: